core/src/main/scala/org/apache/spark/sql/SnappyStrategies.scala

```diff
@@ -134,6 +159,51 @@ private[sql] trait SnappyStrategies {
       case _ => false
     }
 
+    private def getCollocatedPartitioning(leftPlan: LogicalPlan, leftKeys: Seq[Expression],
+        rightPlan: LogicalPlan, rightKeys: Seq[Expression]): (Seq[Expression], Seq[Int], Int) = {
+
+      def getCompatiblePartitioning(plan: LogicalPlan,
+          joinKeys: Seq[Expression]): (Seq[Expression], Seq[Int], Int) = plan match {
+        case PhysicalOperation(_, _, r@LogicalRelation(
+        scan: PartitionedDataSourceScan, _, _)) =>
+          val partCols = scan.partitionColumns.map(colName =>
+            r.resolveQuoted(colName, self.snappySession.sessionState.analyzer.resolver)
+                .getOrElse(throw new AnalysisException(
+                  s"""Cannot resolve column "$colName" among (${r.output})""")))
+          // check if join keys match (or are subset of) partitioning columns
+          val keyOrder = joinKeys.map { k =>
+            val i = partCols.indexWhere(_.semanticEquals(k))
+            if (i < 0) return (Nil, Nil, -1)
+            i
+          }
+          (partCols, keyOrder, scan.numBuckets)
+
+        case PhysicalOperation(_, _, ExtractEquiJoinKeys(_, lKeys, rKeys,
+        _, left, right)) =>
+          // If join is a result of collocated join, then the result can
+          // also be a collocated join with other tables if compatible
+          getCollocatedPartitioning(left, lKeys, right, rKeys)
+
+        case _ => (Nil, Nil, -1)
+      }
+
+      val (leftCols, leftKeyOrder, leftNumPartitions) = getCompatiblePartitioning(
```

Comment:
Check PhysicalOperation.unapply. It breaks the projections and filters already and unwraps the underlying physical operator.